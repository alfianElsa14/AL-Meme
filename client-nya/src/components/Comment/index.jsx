import React, { useState } from 'react'
import { useDispatch } from 'react-redux';
import { useParams } from 'react-router-dom';
import { FormattedMessage } from 'react-intl';

import { calculateTimeDifference } from '@utils/calculateDate';
import { addComment, deleteComment } from '@pages/MemeDetail/actions';
import config from '@config/index';

import DeleteIcon from '@mui/icons-material/Delete';
import { IconButton } from '@mui/material';
import classes from './style.module.scss'

function Comment({ comments, userId }) {
    const dispatch = useDispatch()
    const { id } = useParams()
    const [newComment, setNewComment] = useState('');

    const handleCommentSubmit = (e) => {
        e.preventDefault();
        dispatch(addComment(id, { comment: newComment }));
        setNewComment('');
    };

    const handleDelete = (commentId) => {
        dispatch(deleteComment(commentId, id))
    }
    return (
        <div
            className={classes.comment}
            data-testid="comment"
        >
            <h4><FormattedMessage id='app_comments' /></h4>
            {
                comments?.map((el) => (
                    <div key={el.id} 
                    data-testid="commentList"
                    className={classes.commentList}>
                        <div
                            className={classes.userComment}
                            data-testid="userComment"
                        >
                            <div
                                className={classes.picture}
                                data-testid="picture"

                            >
                                <img src={`${config.api.host}${el.User.imageUrl}`} alt="" />
                            </div>
                            <div className={classes.isiComment}>
                                <p 
                                data-testid="username"
                                className={classes.username}>{el.User.username}</p>
                                <div 
                                className={classes.many}
                                data-testid={`commentText`}
                                >{el.comment}
                                </div>
                            </div>
                        </div>
                        <div className={classes.date}>
                            {
                                (() => {
                                    const { value, unit } = calculateTimeDifference(el.createdAt);
                                    return (
                                        <p>
                                            {value} <FormattedMessage id={`app_${unit}`} /> <FormattedMessage id="app_ago" />
                                        </p>
                                    );
                                })()
                            }
                            {
                                userId === el.User.id && (
                                    <IconButton
                                        onClick={() => handleDelete(el.id)}>
                                        <DeleteIcon className={classes.delete} />
                                    </IconButton>
                                )
                            }
                        </div>
                    </div>
                ))
            }
            <form action="" className={classes.formComment}
                onSubmit={handleCommentSubmit} >
                <textarea
                    className={classes.formText}
                    name=""
                    id=""
                    cols="30"
                    rows="10"
                    value={newComment}
                    onChange={(e) => setNewComment(e.target.value)}
                />
                <button type='submit'><FormattedMessage id="app_comments" /></button>
            </form>
        </div>
    )
}

export default Comment