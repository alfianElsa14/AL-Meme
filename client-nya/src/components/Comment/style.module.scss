@import '@styles/base';

.comment {
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    max-width: 350px;
    width: 100%;
    border: 5px solid $color-violet-dark;
    border-radius: 10px;
    margin-top: 50px;
    padding: 20px;
    background-color: $color-marine-blue;
    margin-bottom: 20px;

    @include apply-at-md-size {
        max-width: 100%;
        width: 50%;
    }

    h4 {
        margin-bottom: 20px;
    }

    .commentList {
        display: flex;
        justify-content: space-between;
        border: 3px solid $color-violet-dark;
        border-radius: 5px;
        margin-bottom: 10px;
        padding: 25px 0;
        width: 100%;


        .userComment {
            display: flex;
            align-items: center;
            padding: 0px 10px;
            gap: 15px;
            overflow: hidden;

            @include apply-at-md-size {
                padding: 0px 30px;
                gap: 30px;
            }

            img {
                width: 23px;
                height: 23px;
                border-radius: 50%;
                border: none;

                @include apply-at-md-size {
                    width: 50px;
                    height: 50px;
                }
            }


            .isiComment {
                display: flex;
                flex-direction: column;
                width: 100%;
                gap: 10px;
                overflow: hidden;

                .username {
                    font-weight: 700;
                    font-size: 13px;
                    margin: 0;

                    @include apply-at-md-size {
                        font-size: 20px;
                    }
                }

                .many {
                    word-wrap: break-word;
                    display: block;
                    font-size: 10px;

                    @include apply-at-md-size {
                        font-size: 15px;
                    }
                }
            }
        }

        .date {
            margin-right: 5px;
            display: flex;
            align-items: center;

            @include apply-at-md-size {
                margin-right: 20px;
            }

            p {
                margin-right: 5px;
                font-size: 5px;

                @include apply-at-md-size {
                    margin-right: 10px;
                    font-size: 13px;
                }
            }

            button {
                border: none;
                background-color: transparent;

                .delete {
                    width: 10px;
                    height: 10px;
                    cursor: pointer;
                    color: red;

                    @include apply-at-md-size {
                        width: 25px;
                        height: 25px;
                        cursor: pointer;
                    }
                }

                .delete:hover {
                    color: rgb(207, 0, 0);
                }
            }
        }
    }

    .formComment {
        display: flex;
        flex-direction: column;
        width: 100%;
        margin-top: 20px;

        .formText {
            resize: none;

            textarea {
                width: 100%;
                height: 50px;

            }
        }

        button {
            margin-top: 20px;
            width: 140px;
            height: 50px;
            border: none;
            border-radius: 5px;
            background-color: rgb(0, 210, 133);
            font-weight: 500;
            cursor: pointer;
        }

        button:hover {
            background-color: rgb(0, 107, 68);
            color: $color-white;
        }
    }
}