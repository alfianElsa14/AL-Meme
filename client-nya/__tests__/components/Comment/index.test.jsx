import Comment from "@components/Comment";
import classes from '@components/Comment/style.module.scss';
import { render } from "@utils/testHelper";

const mockPropsComment = [
    {
        userId,
        memeId,
        comment: 'testt 1123',
        createdAt: new Date(),
        updatedAt: new Date(),
        User: {
            id: 1,
            username: 'mikail',
            email: 'mikail@gmail.com',
            imageUrl: 'image.jpg',
            role: 'basic',
            verified: true,
            verifyToken: null,
            createdAt: new Date(),
            updatedAt: new Date()
        }
    },
]

let wrapper;
let userId = mockPropsComment[0].User.id
let memeId = 1

beforeEach(() => {
    wrapper = render(<Comment comments={mockPropsComment} userId={userId} />)
})

describe('Comment', () => {
    test('Should render Comment correctly', () => {
        const { getByTestId } = wrapper;
        const commentComponent = getByTestId('comment');
        expect(commentComponent).toHaveClass(classes.comment)
        expect(commentComponent).toBeInTheDocument();
    })

    test('Should render Comment list correctly', () => {
        const { getByTestId } = wrapper;
        const commentComponent = getByTestId('commentList');
        expect(commentComponent).toHaveClass(classes.commentList)
        expect(commentComponent).toBeInTheDocument();
    })

    test('Should render picture correctly', () => {
        const { getByTestId } = wrapper;
        const pictureComponent = getByTestId('picture');
        expect(pictureComponent).toHaveClass(classes.picture)
        expect(pictureComponent).toBeInTheDocument();
    })

    test('Should render user comment correctly', () => {
        const { getByTestId } = wrapper;
        const userComment = getByTestId('userComment');
        expect(userComment).toBeInTheDocument();
    });

    test('Should render username correctly', () => {
        const { getByTestId } = wrapper;
        const username = getByTestId('username');
        expect(username).toBeInTheDocument();
    });

    test('Should render comment text correctly', () => {
        const { getByTestId } = wrapper;
        const commentText = getByTestId('commentText');
        expect(commentText).toHaveTextContent('testt 1123');
    });

    test('Should match with snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    })
})