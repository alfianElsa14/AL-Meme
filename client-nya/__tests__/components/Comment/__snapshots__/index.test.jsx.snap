// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comment Should match with snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="comment"
        data-testid="comment"
      >
        <h4>
          Komentar
        </h4>
        <div
          class="commentList"
          data-testid="commentList"
        >
          <div
            class="userComment"
            data-testid="userComment"
          >
            <div
              class="picture"
              data-testid="picture"
            >
              <img
                alt=""
                src="http://localhost:3300/api/image.jpg"
              />
            </div>
            <div
              class="isiComment"
            >
              <p
                class="username"
                data-testid="username"
              >
                mikail
              </p>
              <div
                class="many"
                data-testid="commentText"
              >
                testt 1123
              </div>
            </div>
          </div>
          <div
            class="date"
          >
            <p>
              0
               
              detik
               
              yang lalu
            </p>
            <button
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium delete css-i4bv87-MuiSvgIcon-root"
                data-testid="DeleteIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
        <form
          action=""
          class="formComment"
        >
          <textarea
            class="formText"
            cols="30"
            id=""
            name=""
            rows="10"
          />
          <button
            type="submit"
          >
            Komentar
          </button>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="comment"
      data-testid="comment"
    >
      <h4>
        Komentar
      </h4>
      <div
        class="commentList"
        data-testid="commentList"
      >
        <div
          class="userComment"
          data-testid="userComment"
        >
          <div
            class="picture"
            data-testid="picture"
          >
            <img
              alt=""
              src="http://localhost:3300/api/image.jpg"
            />
          </div>
          <div
            class="isiComment"
          >
            <p
              class="username"
              data-testid="username"
            >
              mikail
            </p>
            <div
              class="many"
              data-testid="commentText"
            >
              testt 1123
            </div>
          </div>
        </div>
        <div
          class="date"
        >
          <p>
            0
             
            detik
             
            yang lalu
          </p>
          <button
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium delete css-i4bv87-MuiSvgIcon-root"
              data-testid="DeleteIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
      <form
        action=""
        class="formComment"
      >
        <textarea
          class="formText"
          cols="30"
          id=""
          name=""
          rows="10"
        />
        <button
          type="submit"
        >
          Komentar
        </button>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
